orttraining/orttraining/core/session/training_session.cc

```diff
@@ -880,6 +891,59 @@ common::Status TrainingSession::GetStateTensors(NameMLValMap& state_tensors) {
   return GetSessionState().GetInitializedTensors(GetStateTensorNames(), allow_missing, state_tensors);
 }
 
+common::Status TrainingSession::GetOptimizerState(std::unordered_map<std::string, NameMLValMap>& opt_state_tensors) {
+  bool allow_missing = (opt_graph_config_.deepspeed_zero.stage != 0);
+  for (const auto& weight_map : weight_to_opt_mapping_) {
+    NameMLValMap curr_opt_tensors;
+    std::unordered_set<std::string> opt_names(weight_map.second.begin(), weight_map.second.end());
+    const auto& weight_name = weight_map.first;
+    GetSessionState().GetInitializedTensors(opt_names, allow_missing, curr_opt_tensors);
+    opt_state_tensors[weight_name] = {};
+    // Keep only prefix in returned value
+    for (const auto& opt_pair: curr_opt_tensors) {
+      const auto& opt_name = opt_pair.first;
+      std::string pre_fix_only = opt_name.substr(0, opt_name.find(weight_name) - 1);
+      opt_state_tensors[weight_name][pre_fix_only] = opt_pair.second;
+    }
+  }
+  return Status::OK();
+}
+
+common::Status TrainingSession::GetModelState(std::unordered_map<std::string, NameMLValMap>& model_state_tensors, bool include_mixed_precision_weights) {
+  bool allow_missing = (opt_graph_config_.deepspeed_zero.stage != 0);
+  std::unordered_set<std::string> fp_tensor_names{};
+  fp_tensor_names.insert(
+      weights_to_train_.begin(), weights_to_train_.end());
```

Comment:
weights_to_train_ doesnt have the updated name of the zero- modified weight. What we must do here is use the updated_weight_names_map_ to also fetch using the updated names, and populate the output using the original weight name. Also, this should be protected by a test for Zero, where we assert that all the weight keys are returned correctly.

Reply:
Changed, and added a test that runs this scenario. Need to design how to assert all the weight keys are returned correctly